x-pack/plugins/canvas/public/lib/aeroelastic/__fixtures__/typescript/typespec_tests.ts

```diff
@@ -0,0 +1,82 @@
+import { select } from '../../select';
+import { Json, Selector } from '../..';
+
+/*
+
+  Type checking isn't too useful if future commits can accidentally weaken the type constraints, because a
+  TypeScript linter will not complain - everything that passed before will continue to pass. The coder
+  will not have feedback that the original intent with the typing got compromised. To declare the intent
+  via passing and failing type checks, test cases are needed, some of which designed to expect a TS pass,
+  some of them to expect a TS complaint. It documents intent for peers too, as type specs are a tough read.
+
+  Run compile-time type specification tests in the `kibana` root with:
+
+     yarn typespec
+
+  Test "cases" expecting to pass TS checks are not annotated, while ones we want TS to complain about
+  are prepended with the comment
+ 
+  // typings:expect-error
+
+  The test "suite" and "cases" are wrapped in IIFEs to prevent linters from complaining about the unused
```

Comment:
What if we used `module.typespec.ts` and told tslint to ignore `.typespec.ts` files? Could we avoid the IIFE gymnastics and make the variable declarations a little more straightforward?

```ts
let plain: Json;
// numbers are OK
plain = 1;
plain = NaN;
plain = Infinity;
plain = -Infinity;
plain = Math.pow(2, 6);
```

Reply:
@spalger thanks for the comments!

Re `module.typespec.ts` I'm not familiar with it, a one-liner PR against this branch or some other pointer would be welcome, or in theory, this conversion could be done as a subsequent PR by your or someone who has the chops.

Re IIFE: my purpose was to 1) make it a bit like `describe` / `it`; 2) use informative function names for some structure without resorting to comment lines, 3) limit the scope of these variables so the reader can see what's what; 4) allow the movement of tested units (now IIFEs) with structured editing as done in Lisp code (whereas if it's line based, the logical boundaries are unclear so it's super error prone to change the sequence of anything or rename something etc.). 

I could solve these with either IIFEs (or analogous) or by using separate files, but I didn't want to make files too small - it's nice to cover a sensible unit in one file (ymmv though). Also, maybe someone spends the time and defines `describe` / `it` functions for making it look even more like unit tests, well assuming it's a goal at all 😄

Reply:
... taking another look, I agree it'd be less tedious if I used eg. `let plain: Json` etc. instead of passing these as typed IIFE args, which indeed complexifies like the WaPo 😸  This is why I stuck with it (welcome to try and suggest alternatives):

- linter will complain that I bind values but don't actually use the value (I could return the value but then the simplicity of returning `void` is gone, so it moves cruft from here to there)
- with the non-value-binding `let`, as there's no initial value, it won't actually adhere to the type unless the type includes the `undefined` value (I'm not sure if it's a tslint issue or if I just found it weird to have a purportedly type X variable which then was `undefined` which is not allowed by the typespec)

One more thing I planned with the functions (which are immediately invoked right now, but don't have to be): potentially reuse them. Eg. pass on various parameters to try things in different combinations.

Reply:
All great points, I'm open to refining the approach in subsequent PRs.